### ISSUE TITLE
"consul watch" CLI command discards argv boundaries

### ISSUE THREAD

**Issue by charles-dyfis-net** [hashicorp/consul#3446]
consul does not pass command-line arguments through faithfully to subprocesses being executed. Instead, it appears to behave in a manner akin to the antipattern `sh -c "$*"`, such that quoting used to control the arguments originally passed to `consul watch` is effectively discarded.

Consider the following (lines starting with `$` are prompt/user input; lines not so prefixed are content written to stdout or stderr):

```
$ ## put a script which prints its arguments, and pipes stdin through jq, in a variable.
$ script=$'printf \'%q\\n\' "$0" "$@"; jq .'
$ 
$ ## demonstrate that this variable was correctly assigned
$ printf '%s\n' "$script"
printf '%q\n' "$0" "$@"; jq .
$ 
$ ## verify that this script works correctly when invoked directly
$ echo '{ "test": "key" }' | bash -xc "$script" _
+ printf '%q\n' _
_
+ jq .
{
  "test": "key"
}
$ 
$ ## verify that this script fails to execute when invoked via ''consul watch''
$ consul watch -type=key -key="/test/key" bash -xc "$script" _
+ printf
printf: usage: printf [-v var] format [arguments]
jq: error: Could not open file _: No such file or directory
Error executing handler: exit status 2
```

---

### `consul version` for both Client and Server

```
Consul 0.9.0
Protocol 2 spoken by default, understands 2 to 3 (agent will automatically use protocol >2 when speaking to compatible agents)
```

...on both ends.

### Operating system and Environment details
Linux/UNIX

**Comment by charles-dyfis-net** [comment]
...reading the source, consul does indeed use `strings.Join(f.Args(), " ")` to whitespace-concatenate arguments, and then calls `agent.ExecScript()`, which calls `sh -c` with that string as a following argument.

Speaking as someone who wears a shell-pedant hat, this is *unfortunate*. It makes passing arbitrary arguments to a command through `consul watch` literally impossible unless one knows the `/bin/sh` implementation in use and can generate a script with that data correctly escaped for its use; and it's liable to make scripts written by people unaware of the relevant caveats vulnerable to shell injection attacks.

**Comment by slackpad** [comment]
Hi @charles-dyfis-net indeed! We've got https://github.com/hashicorp/consul/issues/2999 open, which includes making the shell optional.

**Comment by charles-dyfis-net** [comment]
Even when a shell *is* desired, having the first argument be the script to invoke and subsequent arguments be passed through as... well... *arguments* (which that script can refer to in `$0`, `$1`, etc) is, while perhaps confusing to the user, prone to causing bugs which are less subtle and security-impacting than those caused by present behavior.

That is to say -- what I'd propose as a fix for the case where shell usage *isn't* disabled is to prepend `['sh', '-c']` to the user-provided argument vector. It's a breaking change, to be sure, and we'd want to make it before locking down a 1.0 stable release, but when one is forced to choose between the two, failure cases where code fails outright are better than failure cases where code behaves in subtly wrong ways (and doubly so when those subtly wrong ways can cause data to be parsed as code).

**Comment by charles-dyfis-net** [comment]
I argue that this is not just a usability problem but a security problem. If someone thinks it's safe to run `consul watch /path/to/my/trusted/program "$untrusted_argument"`, and an attacker sets `untrusted_argument=$'$(rm -rf ~)\'$(rm -rf ~)\''`, they're going to have a bad day indeed.

**Comment by slackpad** [comment]
This is fixed in Consul 1.0 via #3509.

Consul agent checks and watches now support an `args` configuration which is a list of arguments to run for the subprocess, which runs the subprocess directly without a shell. The old `script` and `handler` configurations are now deprecated (specify a shell explicitly if you require one). A `-shell=false` option is also available on `consul lock`, `consul watch`, and `consul exec` to run the subprocesses associated with those without a shell.

It seemed like it would cause too much chaos to break this for 1.0, but we've deprecated the right bits and got a path to the safer way, and we intend to remove the shell stuff entirely in a future major release of Consul.
